### PR TITLE
add ipv4ll auto conf

### PR DESCRIPTION
**Pull request by wlan0**
@ibuildthecloud

### CODE REVIEW

File: cmd/network/ipv4ll.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			log.Infof("Set %s on %s", dstIP.String(), link.Attrs().Name)
+			return nil
+		}
+	}
+	log.Error("Could not find a suitable ipv4ll")
+	return fmt.Errorf("Could not find a suitable ipv4ll")
+}
+
+func getNewIPV4LLAddr(randomNum uint32) net.IP {
+	byte1 := randomNum & 255 // use least significant 8 bits
```
Comment (ibuildthecloud) [inline cmd/network/ipv4ll.go]: You need to make sure you never use 169.254.255.255 or 169.254.0.0

Comment (vincent99) [inline cmd/network/ipv4ll.go]: All of 169.254.0.* and 169.254.255.* are not valid:  http://tools.ietf.org/html/rfc3927#section-2.1

File: cmd/network/ipv4ll.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	byte2 := randomNum >> 24 // use most significant 8 bits
+	return []byte{169, 254, byte(byte1), byte(byte2)}
+}
+
+func getPseudoRandomGenerator(ifaceName string) (*rand.Source, error) {
+	haAddr, err := getHaAddr(ifaceName)
```
Comment (ibuildthecloud) [inline cmd/network/ipv4ll.go]: You don't have to jump through so many hoops to do this.  The hardware address is already available to you from the `link.Attr().HardwareAddr`.  The hardware address is really just a `[]byte` so you can just convert that to a in64 with binary.Varint and use that as the seed.

File: cmd/network/ipv4ll.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	if err != nil {
+		log.Error("Could not obtain MAC address")
+		src := rand.NewSource(0)
+		return &src, err
+	}
+	seed := generateSeed(haAddr)
+	fmt.Println(seed)
```
Comment (ibuildthecloud) [inline cmd/network/ipv4ll.go]: don't println

File: cmd/network/ipv4ll.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		dstIP := getNewIPV4LLAddr(randomNum)
+		_, _, err := arping.PingOverIfaceByName(dstIP, ifaceName)
+		if err != nil {
+			// this ip is not being used
+			addr, err := netlink.ParseAddr(dstIP.String() + "/16")
+			if err != nil {
+				log.Errorf("error while parsing ipv4ll addr, err = %v", err)
+				return err
+			}
+			if err := netlink.AddrAdd(link, addr); err != nil {
```
Comment (ibuildthecloud) [inline cmd/network/ipv4ll.go]: What happens if I run this over and over again.  Do I end up assigning a random LL IP each time?  You should probably check to see if a 169.254/16 is already assigned.

Comment (wlan0) [inline cmd/network/ipv4ll.go]: You get the same IP if you run the program again... If you run the for...loop again though, you get a new IP (each time I do `rand.New(*randSource)`)

File: cmd/network/ipv4ll.go (hunk 1 above)
Comment (wlan0) [inline cmd/network/ipv4ll.go]: thanks @vincent99

File: cmd/network/ipv4ll.go (hunk 4 above)
Comment (ibuildthecloud) [inline cmd/network/ipv4ll.go]: If you already have the IP assigned to the NIC the arping will still fail?

Comment (wlan0) [inline cmd/network/ipv4ll.go]: Yes, It fails

File: cmd/network/ipv4ll.go (hunk 2 above)
Comment (wlan0) [inline cmd/network/ipv4ll.go]: When I wrote it initially, I intended it be generic. Once I integrated it with netconf, realized I dont need it to be. I'll change this

**Comment by wlan0** [comment]
@ibuildthecloud updated the code

**Comment by wlan0** [comment]
@ibuildthecloud updated the code

**Comment by ibuildthecloud** [comment]
LGTM.  I'll manually merge this so we can get the godep updates too

**Pull request closed without merging**